lib/drivers/docker/index.js

```diff
@@ -0,0 +1,294 @@
+  switch (req.cmd) {
+    case 'listening':
+      // emit a listening event each time a _new_ port is listened on
+      if (instance.ports.indexOf(req.address.port) < 0) {
+        instance.ports.push(req.address.port);
+        this.emit('listening', instanceId, req.address);
+      }
+  }
+};
+
+DockerDriver.prototype.stopInstance = function(id, style, cb) {
+  debug('DockerDriver.stopInstance(%j, %j, %s)', id, style, cb.name);
+  var instance = this._instance(id);
+  instance.shouldRestart = false;
+  switch (style) {
+    case 'soft':
+      return this.requestOfInstance(id, {cmd: 'stop'}, cb);
+    case 'hard':
+    default:
+      if (instance.current) {
+        return instance.current.kill('SIGTERM', reportExit);
+      } else {
+        return cb(Error('No such instance: ' + id));
+      }
+  }
+  function reportExit(err, res) {
+    debug('stopped container for instance %j:', id, err, res);
+    cb(err, res);
+  }
+};
+
+DockerDriver.prototype._onContainerStopped = function(id, app) {
+  console.error('container stream stopped for some reason!', id, app);
+};
+
+DockerDriver.prototype.dumpInstanceLog = function(id) {
+  debug('DockerDriver.dumpInstanceLog(%j)', id);
+  var instance = this._instance(id);
+  if (!instance.current) {
+    debug('no instance to dump logs for: %j', id);
+    return null;
+  }
+  var logDump = instance.log.duplicate().toString();
+  instance.log.consume(logDump.length);
+  return logDump;
+};
+
+DockerDriver.prototype.updateInstanceEnv = function(id, env, cb) {
+  debug('DockerDriver.updateInstanceEnv(%j)', id);
+  var self = this;
+  var current = this._instance(id).current;
+  if (!current) {
+    debug('no instance to update env of: %j', id);
+    return setImmediate(cb);
+  }
+  // Environment variables can only really have string values, so make sure we
+  // are consistent with that so our comparisons are reliable.
+  env = _.mapValues(env, String);
+  debug('updating instance %j env: %j', id, env);
+  if (_.isEqual(current.env, env)) {
+    debug('env unchanged, skipping restart: %j', id);
+    return cb && setImmediate(cb);
+  } else {
+    debug('env changed, restarting %j: %j => %j', id, current.env, env);
+    // env is alredy a copy
+    current.env = env;
+    return self.requestOfInstance(id, {cmd: 'env-set', env: env}, function() {
+      self.requestOfInstance(id, {cmd: 'restart'}, function(rsp) {
+        cb(rsp && rsp.error, rsp);
+      });
+    });
+  }
+};
+
+DockerDriver.prototype.requestOfInstance = function(id, req, cb) {
+  debug('DockerDriver.requestOfInstance(%j)', id);
+  var current = this._instance(id).current;
+  if (current) {
+    debug('requesting of %j: %j', id, req);
+    return current.request(req, cb);
+  } else if (cb) {
+    setImmediate(cb({error: 'Cannot send request to instance'}));
+  }
+};
+
+DockerDriver.prototype.start = function(instanceMetas, cb) {
+  debug('DockerDriver.start(%j, %s)', instanceMetas, cb.name);
+  // TODO(rmg):
```

Comment:
I added some TODO about this in https://github.com/strongloop-internal/scrum-nodeops/issues/402